### PR TITLE
Support for `Table` output in `Data.read_many`

### LINKED ISSUE #11311: Implement `Data.read_many`
Reads a set of files into a table. The result is either one row per table, or the combined contents of all the tables.

```
Data.read_many
    files:(Vector File)
    format:File_Format=Auto_Detect
    include:Vector File_Attribs=..FileName|..Full_Path|Nothing
    return:Return_As=Merged_Table
    on_problems:Problem_Behavior
    -> Table
```

Implement `File_Format.read_many`: like `File_Format.read` but reads a set of files.

```
File_Format.read_many
    files:(Vector File)
    on_problems:Problem_Behavior
    -> (Vector Any) ! File_Error
```

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #11311 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [ ] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    ## Returns a table with a new column `Value` containing the objects loaded
+       from each file.
+
+       When the source for files to load was a table, all columns from the
+       original table are also retained. In case of name clashes, the newly
+       added columns will get a suffix.
+
+       When the source was a simple Vector, the returned table will also contain
+       a `File Name` column.
+    Table_Of_Objects
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: The name `Table_Of_Objects` was a working name. I'm not sure if we want to keep it or come up with something better?

Definitely `Table_Of_Tables` in Excel sounded much better. But here it may not necessarily be tables, it may be various objects. So `Table_Of_Objects` is quite precise, but it sounds weird.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: Maybe we should change approach altogether and try something like `Added_Column`? (meaning that the loaded files are added as a column to the input)

- `Values_Column`
- `Table_With_New_Column`

any thoughts on these names?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Match_Columns.enso [hunk 2]
```diff
@@ -16,146 +7,3 @@ type Match_Columns
 
        Note: column names are not compared.
     By_Position
-
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Match_Columns.enso:19]: No meaningful code changes in this file, just moving the helpers to separate module - `Match_Columns_Helpers`.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso (hunk 1 above)
Comment (marthasharkey) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: I think `Table_With_New_Column` or `Table_Of_Objects` seem the best to me, `Values_Column` I think could end up being confusing

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: How about more of a boolean? Merged/UnMerged. Merged True/False? I feel like it would be hard for someone to guess what table_of_objects is going to do without reading the help or trying it

Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: I wonder if:

```
type Return_As
   ## Single table with values expanded
   Merged_Table

   ## PRIVATE
       Add this for backwards compatiblity then hide in widget?
   Table_Of_Tables

   ## 
   Individual_Cells

   ## Objects returned as a raw vector allowing zipping to input
   Vector
```

Not sure on name yet either - will think some more

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: > How about more of a boolean? Merged/UnMerged. Merged True/False?

But note that `Merged_Table` has additional arguments (`columns_to_keep` and `match`) that are only relevant in this mode. Switching to a boolean would make this problematic.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: > I wonder if:
> 
> ```
> type Return_As
>    ## Single table with values expanded
>    Merged_Table
> 
>    ## PRIVATE
>        Add this for backwards compatiblity then hide in widget?
>    Table_Of_Tables
> 
>    ## 
>    Individual_Cells
> 
>    ## Objects returned as a raw vector allowing zipping to input
>    Vector
> ```
> 
> Not sure on name yet either - will think some more

`Vector` is handled by `Return_As_Base` - the type needs to be split up so that the `Vector` return works even if `Table` is not imported. It is opaque to the user - the dropdown displays all available types and autoscoping works as shown in tests.

I'd not add `Table_Of_Tables` to `Data.read_many`. It is supported in `Excel_Workbook.read_many` to retain backwards compatibility. But adding it to a brand new function seems unnecessary - why add an obsolete option?

Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: > > How about more of a boolean? Merged/UnMerged. Merged True/False?
> 
> But note that `Merged_Table` has additional arguments (`columns_to_keep` and `match`) that are only relevant in this mode. Switching to a boolean would make this problematic.

Plus dropdown options are clearer in the GUI generally.

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: Yes sorry that wasn't clear. When I said boolean I didn't mean a boolean type. More text options that were Merged or Not Merged.

Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: > `Vector` is handled by `Return_As_Base` - the type needs to be split up so that the `Vector` return works even if `Table` is not imported. It is opaque to the user - the dropdown displays all available types and autoscoping works as shown in tests.
> 
> I'd not add `Table_Of_Tables` to `Data.read_many`. It is supported in `Excel_Workbook.read_many` to retain backwards compatibility. But adding it to a brand new function seems unnecessary - why add an obsolete option?

I was in my head thinking of it as a single type used by both functions but I appreciate we are actually building this dynamically. Just for working out how it might look in the dropdown thinking as a single type is a bit easier!

If the `..Table_Of_Tables` was an option not shown in the dropdown but in one of the *mix in* types then no special handling would be needed for the two reads to share the same object was my thought.

I wonder if:
- `..As_Vector`
- `..With_New_Column` 
- `..As_Merged_Table`

Might be cleanest?

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: > If the `..Table_Of_Tables` was an option not shown in the dropdown but in one of the _mix in_ types then no special handling would be needed for the two reads to share the same object was my thought.

But then more code would be affected by the legacy options, I'd rather not grow places that take legacy options when not needed. Unless there's any problem with the current approach, I'd keep the compatibility conversion as-is.



> I wonder if:
> 
> * `..As_Vector`
> * `..With_New_Column`
> * `..As_Merged_Table`
> 
> Might be cleanest?

That sounds quite clear to me 👍

Comment (AdRiley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: Yes I like those options

**Comment by radeusgd** [comment]
I've got a one remaining failing test:
image

I'm wondering what to do about it - at the moment of `union`, `Sheet Name` is not more special than any other column - the order in which the columns appear depends on the order in which they appear in the inputs - so if the first files are `.tsv` and not `.xls`, it appears after these columns. What do you think? Is that an issue?

Ideally the column should be first _always_. In most cases, when loading only Excel files (or just enough that an `.xls` file goes first) it will already be first. I'm wondering if we should introduce additional logic to handle this edge case where a `.tsv` comes first and possibly changes the ordering.

What's more important I guess is that when matching `By_Position`, the `Sheet Name` column could get merged with the first column of the `.tsv` - that _is_ actually problematic and probably needs for some special logic. I will later add a test for it and try to address it.

-----

Edit: This has been addressed.

I split columns into 3 kinds: input table, metadata and data; each are handled separately. Metadata columns are always matched by name and all are kept (`In_Any`); the matching settings on `As_Matched_Table` only affects the _data_ columns. Tests were added (mostly to `Excel_Spec` as currently only Excel has 'metadata' columns) to illustrate this and check some edge cases.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Return_As_Table.enso]: The options outlined in the comment above have been implemented. `Excel_Workbook.read_many` also allows old options for compatibility, but `Data.read_many` (the new component) only works with new ones.

File: test/Table_Tests/src/IO/Excel_Spec.enso [hunk 3]
```diff
@@ -103,22 +105,42 @@ spec_fmt suite_builder header file read_method sheet_count=5 =
...
-            Problems.test_problem_handling action problems tester ignore_warning_cardinality=True
+            r1 = wb.read_many on_problems=..Report_Error
+            r1.should_fail_with Empty_Sheet
+
+            r2 = wb.read_many on_problems=..Report_Warning
+            r2.row_count . should_equal 25
+            r2.column_names . should_equal ["Sheet Name", "Name", "Quantity", "Price", "A", "B", "C", "D", "E", "Student Name", "Enrolment Date", "Item", "Price 1"]
+            # We also ensure how many rows were loaded from each sheet. There should be no entries for the "Empty" sheet.
+            r2.at "Sheet Name" . to_vector . should_equal <|
+                (Vector.fill 6 "Simple") + (Vector.fill 7 "Strange Dimensions") + (Vector.fill 6 "Dates") + (Vector.fill 6 "Duplicate Columns")
```
Comment (radeusgd) [inline test/Table_Tests/src/IO/Excel_Spec.enso:117]: This check (and analogous one in test much below) is important - due to a bug uncovered by asserts, we were misaligning the names of sheets from their actual content - and thus content fro `Duplicate Columns` was actually assigned `Empty` sheet name - even though Empty sheet did not actually have any rows in the result.

Before the fix, this new test would have failed with:
```
    - [FAILED] should let you read all sheets into a single table [86ms]
        Reason: ['Simple', 'Simple', 'Simple', 'Simple', 'Simple', 'Simple', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Dates', 'Dates', 'Dates', 'Dates', 'Dates', 'Dates', 'Empty', 'Empty', 'Empty', 'Empty', 'Empty', 'Empty'] did not equal ['Simple', 'Simple', 'Simple', 'Simple', 'Simple', 'Simple', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Strange Dimensions', 'Dates', 'Dates', 'Dates', 'Dates', 'Dates', 'Dates', 'Duplicate Columns', 'Duplicate Columns', 'Duplicate Columns', 'Duplicate Columns', 'Duplicate Columns', 'Duplicate Columns']; first difference at index 19  (at X:\NBO\enso\test\Table_Tests\src\IO\Excel_Spec.enso:115-116).
```

We can see `Empty` where 'Duplicate Columns' is expected and then no entries have 'Duplicate Columns' name - the data is messed up and corrupted. Good that this was fixed :)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso [hunk 4]
```diff
@@ -83,6 +83,11 @@ XML_Element.to_table self =
      Otherwise, treated as a single row.
    - fields: a Vector of Text representing the names of fields to look up.
      If `Nothing` then all fields found are added.
+   - treat_dictionary_as_sequence: If set to `True`, then a dictionary (and
```
Comment (GregoryTravis) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso]: If this applies to dictionaries and other types as well, perhaps `treat_as_sequence` would be more precise.

File: test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso [hunk 5]
```diff
@@ -519,6 +530,37 @@ run_union_tests group_builder setup call_union =
...
+        t3 = table_builder [["A", [Nothing, Nothing]]]
+        t4 = table_builder [["A", ['a', 'b']]]
+        t5 = table_builder [["A", [Nothing]]]
+
+        t1.at "A" . value_type . should_equal Value_Type.Mixed
+        setup.expect_integer_type <| t2.at "A"
+
+        t = call_union [t1, t2, t3, t4, t5]
+        expect_column_names ["A"] t
+        t.at "A" . to_vector . should_equal [Nothing, '2', '3', Nothing, 'a', 'b']
```
Comment (GregoryTravis) [inline test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso]: Why does the `Nothing` in `t5` not show up at the end of this list?

File: test/Table_Tests/data/transient/sqlite_test.db-journal
Comment (jdunkerley) [inline test/Table_Tests/data/transient/sqlite_test.db-journal:1]: shouldn't be committed.

File: distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso [hunk 6]
```diff
@@ -2460,19 +2460,25 @@ type DB_Table
        - at_least_one_row: for an empty aggregate value, if `at_least_one_row` is
          true, a single row is output with `Nothing` for the aggregates column; if
          false, no row is output at all.
+       - sequences_only: if set to true, only sequence-like values are expanded.
+         Otherwise (default), the operation tries to expand also other kinds of
+         aggregates, e.g. dictionaries.
```
Comment (jdunkerley) [inline distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso]: Think ok for now but think the text here needs some more thought.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso [hunk 7]
```diff
@@ -375,17 +386,71 @@ build_sheet_selector workbook:Excel_Workbook display:Display=Display.Always -> W
...
+   This type is deprecated and has been superseded by `Return_As_Table`.
+   It is kept only for compatibility reasons.
 type Return_As
-    ## Each sheet is returned as a row.
+    ## DEPRECATED
+       Each sheet is returned as a row.
     Table_Of_Tables
 
     ## All sheets are merged into a single table. A union operation is performed.
     Merged_Table (columns_to_keep : Columns_To_Keep = Columns_To_Keep.In_Any) (match : Match_Columns = Match_Columns.By_Name)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso]: should be DEPRECATED as well

File: distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso [hunk 8]
```diff
@@ -375,17 +386,71 @@ build_sheet_selector workbook:Excel_Workbook display:Display=Display.Always -> W
...
 
-## PRIVATE
-   Creates a query widget for the `read` method.
+## A helper method that ensures that we can take the new `Return_As_Table` type
+   as well as the deprecated `Return_As` type, together with autoscoping.
+   With autoscoping, the `As_Merged_Table` constructor is ambiguous, so plain
+   autoscoping would not work and we do some workarounds to make it work.
+   In case of the ambiguity, the new-style `Return_As_Table.As_Merged_Table`
+   variant is preferred.
+private _resolve_return_as_compatibility return =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Excel/Excel_Workbook.enso:416]: We should attach a deprecated warning to the result about the argument.

**Comment by AdRiley** [comment]
image

Are we able to make this a vector of file pickers? Don't know if the widget system allows this?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso (hunk 4 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso]: I think it is only dictionary like objects not sure we have the naming quite right.

I suggest for now we don't include it on `Table.from_objects` lets think on the API design a little.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso:1]: Following our code standard discussion should this be in Helper namespace? 
Though I guess we should do a whole refactor in a single pass.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso (hunk 4 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso]: Perhaps, but I don't think such name is good. Regardless of the setting, we always treat e.g. a `Table` or `Column` as a sequence. It would be misleading to have `treat_as_sequence=False` and then most types are still treated as sequence...

This setting deals only with the less-sequence-like types, mostly dictionary-like. Perhaps `try_treating_all_types_as_sequence` could be a more precise name, what do you think? I'm worried it will be a bit too scary for users though... This setting overall is a bit complex. Maybe we shouldn't expose it publicly after all?

**Comment by AdRiley** [comment]
image

Is this a defect or user error?

**Comment by AdRiley** [comment]
image

Who are you calling a dummy? ;)

**Comment by radeusgd** [comment]
> image
> 
> Is this a defect or user error?

Looks like a defect I guess, but I'm not exactly sure what I'm seeing? What is happening here that the 'not exists' warning may be reported?

**Comment by AdRiley** [comment]
> > image
> > Is this a defect or user error?
> 
> Looks like a defect I guess, but I'm not exactly sure what I'm seeing? What is happening here that the 'not exists' warning may be reported?

I think it was because I didn't choose a format

**Comment by radeusgd** [comment]
> image
> 
> Who are you calling a dummy? ;)

Ooops, thanks for spotting that! This is an internal column that is used to help with merging empty tables to the result. I thought I had a test that checked that it doesn't "escape", thanks for finding an edge case I missed.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data.enso [hunk 9]
```diff
@@ -140,12 +140,13 @@ read path=(Missing_Argument.throw "path") format=Auto_Detect (on_problems : Prob
          example_csv_dir_to_table = Data.read_many files
 @paths (Vector_Editor item_editor=Text_Input item_default='""' display=..Always)
 @format File_Format.default_widget
+@return Return_As.default_widget
 read_many : Many_Files_List -> File_Format -> Return_As -> Problem_Behavior -> Any ! File_Error
-read_many (paths : Many_Files_List = Missing_Argument.throw "paths") format=Auto_Detect return=..Vector (on_problems : Problem_Behavior = ..Report_Warning) =
+read_many (paths : Many_Files_List = Missing_Argument.throw "paths") format=Auto_Detect return=(Return_As.default_value paths) (on_problems : Problem_Behavior = ..Report_Warning) =
```
Comment (AdRiley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data.enso:145]:
```suggestion
read_many (paths : Many_Files_List = Missing_Argument.throw "paths") format=..Auto_Detect return=(Return_As.default_value paths) (on_problems : Problem_Behavior = ..Report_Warning) =
```

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Read_Many_Helpers.enso [hunk 10]
```diff
@@ -28,3 +33,83 @@ ensure_column_type_valid_to_be_files_list (column : Column) ~action =
...
+    Runtime.assert (empty_table.row_count == 0) "The table must be empty."
+    Table.new (empty_table.columns.map c-> Column.from_vector c.name [Nothing] c.value_type)
+
+_add_objects_column (base_table : Table) (objects : Vector Any) -> Table  =
+    unique_naming = base_table.column_naming_helper.create_unique_name_strategy
+    unique_naming.mark_used base_table.column_names
+    objects_column_name = unique_naming.make_unique "Value"
+    base_table.set (Column.from_vector objects_column_name objects) as=objects_column_name set_mode=..Add
+
+## Workaround for bug https://github.com/enso-org/enso/issues/11570
```
Comment (jdunkerley) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Read_Many_Helpers.enso:111]: Should we have `TODO: Remove workaround one #11570 is closed.`

File: test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso (hunk 5 above)
Comment (jdunkerley) [reply test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso]: Why isn't the result `[Nothing, '2', '3', Nothing, Nothing, 'a', 'b', Nothing]` ?

File: test/Table_Tests/src/IO/Excel_Spec.enso [hunk 11]
```diff
@@ -103,22 +105,42 @@ spec_fmt suite_builder header file read_method sheet_count=5 =
...
+            w1 = Problems.expect_warning No_Rows r2
+            w1.to_display_text . should_contain "The sheet Empty failed to load"
+            w1.to_display_text . should_contain "There is no data in the sheet."
+            w2 = Problems.expect_warning Duplicate_Output_Column_Names r2
+            w2.column_names . should_equal ["Price"]
 
         group_builder.specify "should let you read all sheets into a table of tables" <|
             wb = read_method file
-            action = wb.read_many return=..Table_Of_Tables on_problems=_
+            action = wb.read_many return=..With_New_Column on_problems=_
```
Comment (jdunkerley) [inline test/Table_Tests/src/IO/Excel_Spec.enso:126]: We should keep a deprecrated test too,

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso [hunk 12]
```diff
@@ -19,4 +21,4 @@ type Many_Files_List
 
 ## PRIVATE
 Many_Files_List.from (that : Vector) =
-    Many_Files_List.Value that that
+    Many_Files_List.Value that that Return_As_Base.As_Vector
```
Comment (AdRiley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: I think the default for files should be As_Merged_Table

**Review by jdunkerley — CHANGES_REQUESTED** [review]
I think we should remove the dictionary support from the API at the moment.
It was added for some work with Ken and we need to do some design on the API.

**Comment by jdunkerley** [comment]
> > image
> > Is this a defect or user error?
> 
> Looks like a defect I guess, but I'm not exactly sure what I'm seeing? What is happening here that the 'not exists' warning may be reported?

Think this is probably the visualization error the GUI is looking at.

File: test/Table_Tests/src/IO/Excel_Spec.enso (hunk 11 above)
Comment (radeusgd) [reply test/Table_Tests/src/IO/Excel_Spec.enso:126]: There is one just below. I'll add a warning check as well.

**Comment by AdRiley** [comment]
image

Trying to use ..In_List gives bad error message.

Then the widget is broken if I try to add column_names

image

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso (hunk 12 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: We cannot do that, because `As_Merged_Table` is only available in `Table` library. This has to work with just the `Base`.

_Technically_, we could make it so that _once `Table` is imported_, we change from `As_Vector` to `As_Merged_Table`. But I **really** don't like the concept that adding an import makes our code change behaviour in such ways...

(Well we do change behaviour as e.g. adding a `Table` import will make `Auto_Detect` able to parse CSV, whereas without that import it would have failed - but IMO change from 'fails' to 'succeeds' is ok, but from one return type to another is problematic. Open to discussion though.)

**Comment by jdunkerley** [comment]
> image
> 
> Are we able to make this a vector of file pickers? Don't know if the widget system allows this?

Yes should be do-able.
`@path (Widget.Vector_Editor item_editor=Widget.File_Browse item_default='""' display=..Always)`

**Comment by jdunkerley** [comment]
> image
> 
> Trying to use ..In_List gives bad error message.
> 
> Then the widget is broken if I try to add column_names
> 
> image

We need to add a missing required error here so that the auto-scoping works,
Its a limitation of the autoscoping set up.

**Comment by AdRiley** [comment]
Maybe out of scope for this MR, but this warning is very long

image

Warning: No common type was found for types: Char (variable length, max_size=unlimited), Char (variable length, max_size=unlimited), Char (variable length, max_size=unlimited), Integer (64 bits) when unifying column [Sales], so the values were converted to text. If you want to have mixed types instead, please cast one of the columns to `Mixed` beforehand.

Wonder if we should 
1. Remove the duplicate types
2. Remove the last sentence about Mixed

So then 

Warning: No common type was found for types: Char (variable length, max_size=unlimited), Integer (64 bits) when unifying column [Sales], so the values were converted to text.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso (hunk 12 above)
Comment (AdRiley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: With my product manager hat on. This has to be the default for when I am reading in multiple Excel workbooks. The other options just don't make any sense in the GUI.

With my engineering hat on, I'm not entirely sure how we solve this.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: I see.

1. What is your opinion on changing the default behaviour based on the imports? 
Engineering-wise that can surely be implemented quite easily, I was just hesitant as I don't like such behaviour change... But I don't have many better ideas. In GUI the `Table` will be imported 99% of the time, so I guess the behaviour will be fairly consistent, we can document it.

2. _Alternatively_, we could make `return` a required argument (throw the `Missing_Argument` error) and then the user would choose explicitly every time. We can even make `As_Merged_Table` the 'top' result in dropdown so it is easily accessible. Although I feel like having to set this argument every time instead of having a default will be quite annoying...

3. Third option is to keep the behaviour change less disruptive. As noted above - I think behaviour change that goes from 'error' to 'working' when an import is added (e.g. like with new supported formats in `Auto_Detect` made available by imports) is acceptable - it is much less disruptive than the type of result in unexpected ways. So we could make it so that the default is `As_Merged_Table` but if `Table` is not imported it raises an error:
    > Table output was selected, but the Table library is not imported. Please either import `Standard.Table` or change the return type to `As_Vector`.

-------

I feel like (2) is nicely consistent but too annoying in our most usual case.

I like (3) the most. I feel like it is the best compromise we can get. In the most popular use case (GUI with Table basically always imported) it works correctly and does not require unnecessary setting of arguments. In rare cases in GUI where the import is missing, it gives a clear error that tells the user possible resolutions - either add the import or explicitly request `Vector` return type.

File: test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso (hunk 5 above)
Comment (radeusgd) [reply test/Table_Tests/src/Common_Table_Operations/Join/Union_Spec.enso]: I think because I've made a mistake and haven't run this test to find out 😅 

The test is marked as pending, to be resolved with #6281. So I would have probably caught it when implementing that, but ofc good to catch early - I will amend the test - thanks for catching it.

File: test/Table_Tests/data/transient/sqlite_test.db-journal
Comment (radeusgd) [reply test/Table_Tests/data/transient/sqlite_test.db-journal:1]: I agree it shouldn't - that's why I have removed it.

It was originally added in another PR in the past: https://github.com/enso-org/enso/commit/676a7d4256a892e7b7772d73f0ab6c20ea8491d0#diff-6a61cf5bec88d85f6ebb57b94209d2a9bb1aeb4bf0b4d13db8527ae282dbdfa3

Comment (radeusgd) [reply test/Table_Tests/data/transient/sqlite_test.db-journal:1]: I have also updated `.gitignore` so that it should be far less likely to get such accidental commits in the future.

Previously, `.gitignore` was ignoring only some known file extensions (and this journal was not something we've expected). Now, all files in the `transient` directory are ignored by default.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso:1]: Hmm, tbh I don't remember exactly. Could you please post the notes doc that we've got from the discussions somewhere? I don't remember all the guidelines we agreed on so would be good to have a place to refer to.

File: distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso (hunk 6 above)
Comment (radeusgd) [reply distribution/lib/Standard/Database/0.0.0-dev/src/DB_Table.enso]: Removing the public API for now, it will only be kept internally.

Added ticket #11681 to design the public API.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Match_Columns_Helpers.enso:1]: Putting them together

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso (hunk 12 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: We should default to merged table.

If we made the default `..As_Merged_Table` code, the resolution would be late and so would work if the Table is loaded into scope.

We can add a special case in the resolve handler so if not resolved it tries to resolve something which would let us throw and error saying missing the table input.

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Many_Files_List.enso]: After discussion it seems that (3) is what we want

File: distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso (hunk 4 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso]: I see your point, `treat_dictionary_as_sequence` is good.

**Review by GregoryTravis — APPROVED** [review]

**Comment by radeusgd** [comment]
- Updating the widget to be a vector of file pickers.
- Fixing `In_List` to throw missing argument.
image

**Comment by radeusgd** [comment]
> Maybe out of scope for this MR, but this warning is very long
> 
> image
> 
> Warning: No common type was found for types: Char (variable length, max_size=unlimited), Char (variable length, max_size=unlimited), Char (variable length, max_size=unlimited), Integer (64 bits) when unifying column [Sales], so the values were converted to text. If you want to have mixed types instead, please cast one of the columns to `Mixed` beforehand.
> 
> Wonder if we should
> 
> 1. Remove the duplicate types
> 2. Remove the last sentence about Mixed
> 
> So then
> 
> Warning: No common type was found for types: Char (variable length, max_size=unlimited), Integer (64 bits) when unifying column [Sales], so the values were converted to text.

Sure that's an easy change - just added `distinct` and removed the suffix, now we have:
image

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Return_As.enso [hunk 13]
```diff
@@ -32,16 +33,25 @@ type Return_As
...
+        self.underlying.make_return input objects on_problems
 
     ## PRIVATE
        Resolve an unresolved constructor to the actual type.
     private resolve value = case value of
         _ : Function ->
+            # Result when we fail to resolve an autoscoped constructor.
+            symbol_not_found =
+                # TODO ideally we'd like to use `Meta` but it currently has no support for UnresolvedConstructor so we use a 'hack'.
+                is_the_default = value.to_text == "..As_Merged_Table"
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Return_As.enso]: Checking `to_text` feels like a bit of a hack, but currently `Meta` seems to lack support for `UnresolvedConstructor` - `Meta.meta` returns `Primitive` and does not allow any inspecting.

It feels like it may prove useful in the future, @jdunkerley do we want to create a ticket for the engine team to add support for `Unresolved_Constructor` in `Meta`?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso (hunk 4 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Extensions/Table_Conversions.enso]: The argument was removed from public API for now, will be reintroduced with new design as part of #11681

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Return_As.enso (hunk 13 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Read/Return_As.enso]: I would have done it via
```
type My_Private_Type
    As_Merged_Table
```
and trying to convert to that.

I agree we could probably do with something on Meta.
cc: @JaroslavTulach  for visibility.

**Review by jdunkerley — APPROVED** [review]
